pkg/hnsw/vector.go

```diff
@@ -0,0 +1,122 @@
+package hnsw
+
+import (
+	"errors"
+	"math"
+)
+
+type Point []float32
+
+type Vector struct {
+	id    Id
+	point Point
+
+	friends []*BaseQueue
+}
```

Comment:
You might be able to simplify this struct even further by instead caching the distances for each adjacency in the queue itself.

Reply:
> You might be able to simplify this struct even further by instead caching the distances for each adjacency in the queue itself.

Sorry, I'm a bit confused, what do you mean by caching?

Reply:
Oh, it actually does track the distance here: https://github.com/kevmo314/appendable/blob/3aeeec2039166c2b11f59214c4e349716fef57e7/pkg/hnsw/eucqueue.go#L47

Is it necessary to have the euclidean distance functions on the vector struct in that case? It might be easier to pass in the distance as an argument.

Reply:
> Is it necessary to have the euclidean distance functions on the vector struct in that case?

IMO, yes because we compute these distances on the fly relative to the vector struct's `Point`. Euclidian distance-wise, I figured it would be nice to use a `func (v *Vector)`.

Reply:
Right, but we don't need to recompute them all the time because they're already stored in the queue? So the distance is only needed at insert time.

Reply:
Oh I see. Yeah, I updated the PR so that the friend insertion has the following header:

```go
func (v *Vector) InsertFriendsAtLevel(level int, friendId Id, dist float32) {
```